File: Geo-computing/python/lab0/dsp.py
#!/usr/bin/env python


def smooth (a, x, y):
    n = len(x) 
    b = 1.0-a 
    yi = y[0] = x[0]

    for i in range(1,n-1,1):
        y[i] = yi = a*yi + b*x[i] 

    y[n-1] = yi = (a*yi +x[n-1])/(1.0+a) 
    for i in range(n-2,-1,-1):
        y[i] = yi = a*yi + b*y[i]

def smooth2 (a, x, y):
    n = len(x) 
    b = 1.0-a
    sx = b ; sy = a 
    yi = y[0] = sx*x[0]

    for i in range(1,n-1,1):
        y[i] = yi = a*yi + b*x[i] 

    sx /= 1.0+a
    sy /= 1.0+a
    y[n-1] = yi = sy*yi+sx*x[n-1];

#    y[n-1] = yi = (a*yi +x[n-1])/(1.0+a) 
    for i in range(n-2,-1,-1):
        y[i] = yi = a*yi + b*y[i]

def mean(x):
    n = len(x) 
    return sum(x)/n

File: Geo-computing/java/lab0/Dsp.java
package lab0;

/**
 * Digital signal processing.
 * @author Dave Hale, Colorado School of Mines
 * @version 2012.08.19
 */
public class Dsp {
  public static void smooth(float a, float[] x, float[] y) {
    int n = x.length;
    float b = 1.0f-a;
    float yi = y[0] = x[0]; // first
    for (int i=1; i<n-1; ++i) // forward
      y[i] = yi = a*yi+b*x[i];
    y[n-1] = yi = (a*yi+x[n-1])/(1.0f+a); // last
    for (int i=n-2; i>=0; --i) // reverse
      y[i] = yi = a*yi+b*y[i];
  }
  public static float mean(float[] x) {
    int n = x.length;
    float sum = 0.0f;
    for (int i=0; i<n; ++i)
      sum += x[i];
    return sum/n;
  }
}
